test/d2i_test.c

```diff
@@ -69,36 +69,37 @@ static int test_bad_asn1()
      * decoder is called.
      */
     len = BIO_read(bio, buf, sizeof buf);
-    if (len < 0)
+    if (!TEST_int_ge(len, 0))
         goto err;
 
     value = ASN1_item_d2i(NULL, &buf_ptr, len, item_type);
     if (value == NULL) {
-        if (expected_error == ASN1_DECODE)
+        if (TEST_int_eq((int)expected_error, (int)ASN1_DECODE))
             ret = 1;
         goto err;
     }
 
     derlen = ASN1_item_i2d(value, &der, item_type);
 
     if (der == NULL || derlen < 0) {
-        if (expected_error == ASN1_ENCODE)
+        if (TEST_int_eq((int)expected_error, (int)ASN1_ENCODE))
             ret = 1;
         goto err;
     }
 
     if (derlen != len || memcmp(der, buf, derlen) != 0) {
-        if (expected_error == ASN1_COMPARE)
+        if (TEST_int_eq((int)expected_error, (int)ASN1_COMPARE))
             ret = 1;
         goto err;
     }
 
-    if (expected_error == ASN1_OK)
+    if (TEST_int_eq((int)expected_error, (int)ASN1_OK))
         ret = 1;
 
  err:
     /* Don't indicate success for memory allocation errors */
-    if (ret == 1 && ERR_GET_REASON(ERR_peek_error()) == ERR_R_MALLOC_FAILURE)
+    if (ret == 1 && !TEST_false(ERR_GET_REASON(ERR_peek_error())
+                                == ERR_R_MALLOC_FAILURE))
```

Comment:
Style wise, I think this would be better:

``` C
    if (ret == 1
        && !TEST_false(ERR_GET_REASON(ERR_peek_error()) == ERR_R_MALLOC_FAILURE))
```